spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/security/oauth2/client/OAuth2ClientAutoConfiguration.java

```diff
@@ -0,0 +1,182 @@
+import org.springframework.security.oauth2.client.registration.ClientRegistrationProperties;
+import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository;
+import org.springframework.security.oauth2.client.registration.InMemoryClientRegistrationRepository;
+
+/**
+ * {@link EnableAutoConfiguration Auto-configuration} for Spring Security OAuth 2.0 / OpenID Connect 1.0 client support.
+ *
+ * @author Joe Grandja
+ * @since 2.0.0
+ */
+@Configuration
+@ConditionalOnWebApplication(type = Type.SERVLET)
+@ConditionalOnClass({EnableWebSecurity.class, ClientRegistration.class})
+@AutoConfigureBefore(SecurityAutoConfiguration.class)
+public class OAuth2ClientAutoConfiguration {
+
+	private static final String CLIENT_PROPERTY_PREFIX = "security.oauth2.client";
+
+	private static final String CLIENT_ID_PROPERTY = "client-id";
+
+	private static final String USER_INFO_URI_PROPERTY = "user-info-uri";
+
+	private static final String USER_NAME_ATTR_NAME_PROPERTY = "user-name-attribute-name";
+
+	static Set<String> getClientKeys(Environment environment) {
+		return getClientPropertiesByClient(environment).keySet();
+	}
+
+	static Map<String, Map> getClientPropertiesByClient(Environment environment) {
+		Map<String, Object> clientPropertiesByClient = Binder.get(environment)
+				.bind(CLIENT_PROPERTY_PREFIX, Bindable.mapOf(String.class, Object.class))
+				.orElse(new HashMap<>());
+
+		// Filter out clients that don't have the client-id property set
+		return clientPropertiesByClient.entrySet().stream()
+				.map(e -> new AbstractMap.SimpleImmutableEntry<>(e.getKey(), (Map) e.getValue()))
+				.filter(e -> e.getValue().containsKey(CLIENT_ID_PROPERTY))
+				.collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+	}
+
+	@Order(1)
+	@Configuration
+	@Conditional(ClientsConfiguredCondition.class)
+	@ConditionalOnMissingBean(ClientRegistrationRepository.class)
+	protected static class ClientRegistrationRepositoryConfiguration {
+		private final Environment environment;
+
+		protected ClientRegistrationRepositoryConfiguration(Environment environment) {
+			this.environment = environment;
+		}
+
+		@Bean
+		public ClientRegistrationRepository clientRegistrationRepository() {
+			List<ClientRegistration> clientRegistrations = new ArrayList<>();
+
+			Binder binder = Binder.get(this.environment);
+			getClientKeys(this.environment).forEach(clientKey -> {
+				String fullClientKey = CLIENT_PROPERTY_PREFIX + "." + clientKey;
+				ClientRegistrationProperties clientRegistrationProperties = binder.bind(
+						fullClientKey, Bindable.of(ClientRegistrationProperties.class)).get();
+				clientRegistrations.add(new ClientRegistration.Builder(clientRegistrationProperties).build());
+			});
+
+			return new InMemoryClientRegistrationRepository(clientRegistrations);
+		}
+	}
+
+	@Order(2)
+	@ConditionalOnMissingBean(WebSecurityConfiguration.class)
+	@ConditionalOnBean(ClientRegistrationRepository.class)
+	@EnableWebSecurity
```

Comment:
I don't think this should be annotated with `@EnableWebSecurity`. IIRC, we've never needed it in multiple places in other Spring Boot security configuration classes. I think we just need to make sure that the auto-configuration class that enables web security is doing its job at the right time relative to these configuration classes.

Reply:
Agreed. Will replace `@EnableWebSecurity` with `@Configuration`